CDP4Dal/Session.cs

```diff
@@ -491,17 +500,21 @@ public async Task Read(Thing thing, IQueryAttributes queryAttributes)
             var dto = thing.ToDto();
 
             // Create the token source
-            this.cancellationTokenSource = new CancellationTokenSource();
+            var cancellationTokenSource = new CancellationTokenSource();
+            var cancellationTokenKey = Guid.NewGuid();
+            this.cancellationTokenSourceDictionary.TryAdd(cancellationTokenKey, cancellationTokenSource);
+
             IEnumerable<CDP4Common.DTO.Thing> dtoThings;
 
             try
             {
-                dtoThings = await this.Dal.Read(dto, this.cancellationTokenSource.Token, queryAttributes);
+                dtoThings = await this.Dal.Read(dto, cancellationTokenSource.Token, queryAttributes);
+                this.cancellationTokenSourceDictionary.TryRemove(cancellationTokenKey, out cancellationTokenSource);
             }
             catch (OperationCanceledException)
             {
                 logger.Info("Session.Read {0} {1} cancelled", thing.ClassKind, thing.Iid);
-                this.cancellationTokenSource = null;
+                this.cancellationTokenSourceDictionary.TryRemove(cancellationTokenKey, out _);

```

Comment:
please use the same pattern as in the `Read` method where we have a 
```
try catch finally
```

where in the finally we should have the `TryRemove` on the `cancellationTokenSourceDictionary` and i would prefer syntax is the same all over and explicit and easy to read.